_examples/render/main.go

```diff
@@ -0,0 +1,153 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"reflect"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/pressly/chi"
+	"github.com/pressly/chi/middleware"
+	"github.com/pressly/chi/render"
+)
+
+type RuntimeObject struct {
+	ID       int      `json:"id"`
+	AuthData string   `json:"auth_data"`
+	Data     []string `json:"data"`
+}
+
+type PresenterObject struct {
+	*RuntimeObject `json:",inline"`
+
+	URL   string `json:"url"`
+	False bool   `json:"false"`
+
+	// Omit by default. Show explicitly for auth'd users only.
+	AuthData interface{} `json:"auth_data,omitempty"`
+}
+
+type PresenterObjectV2 struct {
+	*PresenterObject `json:",inline"`
+
+	ResourceURL string `json:"resource_url"`
+
+	// Omit.
+	URL interface{} `json:"url,omitempty"`
+}
+
+type PresenterObjectV1 struct {
+	*PresenterObjectV2 `json:",inline"`
+
+	Data map[string]bool `json:"data"`
+}
+
+var (
+	ErrUnauthorized = errors.New("Unauthorized")
+	ErrForbidden    = errors.New("Forbidden")
+	ErrNotFound     = errors.New("Resource not found")
+
+	v2 = render.NewPresenter()
+	v1 = render.NewPresenter()
+)
+
+func init() {
+	render.Respond = customRespond
+
+	render.DefaultPresenter.Register(func(ctx context.Context, from *RuntimeObject) (*PresenterObject, error) {
+		to := &PresenterObject{
+			RuntimeObject: from,
+			URL:           fmt.Sprintf("https://api.example.com/objects/%v", from.ID),
+		}
+		// Only show to auth'd user.
+		if _, ok := ctx.Value("auth").(bool); ok {
+			to.AuthData = from.AuthData
+		}
+		return to, nil
+	})
+
+	v2.RegisterFrom(render.DefaultPresenter)
+	v2.Register(func(ctx context.Context, from *PresenterObject) (*PresenterObjectV2, error) {
+		return &PresenterObjectV2{PresenterObject: from, ResourceURL: from.URL}, nil
+	})
+
+	v1.RegisterFrom(v2)
+	v1.Register(func(ctx context.Context, from *PresenterObjectV2) (*PresenterObjectV1, error) {
+		to := &PresenterObjectV1{
+			PresenterObjectV2: from,
+			Data:              map[string]bool{},
+		}
+		for _, item := range from.Data {
+			to.Data[item] = true
+		}
+		return to, nil
+	})
+}
+
+// customRespond sets response status code based on Error value/type.
+func customRespond(ctx context.Context, w http.ResponseWriter, v interface{}) {
+	val := reflect.ValueOf(v)
+	if err, ok := val.Interface().(error); ok {
+		switch err {
+		case ErrUnauthorized:
+			ctx = render.Status(ctx, 401)
+		case ErrForbidden:
+			ctx = render.Status(ctx, 403)
+		case ErrNotFound:
+			ctx = render.Status(ctx, 404)
+		default:
+			ctx = render.Status(ctx, 500)
+		}
+		render.DefaultRespond(ctx, w, map[string]string{"error": err.Error()})
+		return
+	}
+
+	render.DefaultRespond(ctx, w, v)
+}
+
+func main() {
+	r := chi.NewRouter()
+
+	r.Use(middleware.RequestID)
+	r.Use(middleware.Logger)
+	r.Use(middleware.Recoverer)
+	r.Use(render.ParseContentType)
```

Comment:
we may want to support render.Status(200) or something, so we can override the status code for a response. the http.ResponseWriter has w.WriteHeader(statusInt), but, this will write the header right then, which isn't what we want.

the consideration is, what should be the signature of render.Respond(), I do believe we want to give control

perhaps we have.. render.Respond(w, r, Object) and render.RespondStatus(w, r, status, Object)